nitro-src/claim.ts

```diff
@@ -1,163 +1,219 @@
     // Any guarantees before this one have priority on the funds
     // So we must account for that by reducing the surplus
-    for (
-      let guaranteeIndex = 0;
-      guaranteeIndex < targetChannelIndex;
-      guaranteeIndex++
-    ) {
-      const { amount } = guarantees[guaranteeIndex];
+    const guaranteesBeforeOurs = guaranteesForOneAsset.slice(
+      0,
+      targetChannelIndex
+    );
+    const amountBeforeUs = sumAllocationAmounts(guaranteesBeforeOurs);
+    let maxAmountCanPayOut = holdings[assetIndex].sub(amountBeforeUs);
 
-      surplus = surplus.sub(min(BigNumber.from(amount), surplus)); // Prevent going below 0
-    }
-    // If there are not enough funds to fund the guarantee we return immediately
-    if (surplus.isZero()) {
-      return {
-        updatedGuaranteeOutcome,
-        updatedHoldings,
-        updatedTargetOutcome,
-        exit,
-      };
+    // If there are not enough funds to fund the guarantee, we skip the asset
+    if (maxAmountCanPayOut.lte(0)) {
+      continue;
     }
 
-    surplus = min(
-      surplus,
-      BigNumber.from(guarantees[targetChannelIndex].amount)
+    maxAmountCanPayOut = min(
+      maxAmountCanPayOut,
+      BigNumber.from(guaranteesForOneAsset[targetChannelIndex].amount)
     );
 
-    let exitRequestIndex = 0;
+    const {
+      afterClaimAllocations,
+      afterClaimGuaranteeAmount,
+      newExits,
+    } = claimOneGuaranteeForOneAsset(
+      guaranteesForOneAsset[targetChannelIndex],
+      targetAllocations,
+      maxAmountCanPayOut,
+      exitRequest[assetIndex]
+    );
 
-    const destinations = decodeGuaranteeData(
-      guarantees[targetChannelIndex].metadata
+    // How much did all of the exits pay out?
+    const amountPaidOut = sumAllocationAmounts(newExits);
+    // For an asset, update funds based on payout
+    afterClaimHoldings[assetIndex] = afterClaimHoldings[assetIndex].sub(
+      amountPaidOut
     );
-    // Iterate through every destination in the guarantee's destinations
+    // Update the guarantee to account for the payout
+    afterClaimGuarantee[assetIndex].allocations[
+      targetChannelIndex
+    ].amount = afterClaimGuaranteeAmount.toHexString();
+
+    afterClaimTargetOutcome.push({
+      asset: targetOutcome[assetIndex].asset,
+      metadata: targetOutcome[assetIndex].metadata,
+      allocations: afterClaimAllocations,
+    });
+
+    singleAssetExit.allocations = newExits;
+    afterClaimExits.push(singleAssetExit);
+  }
+
+  return {
+    updatedHoldings: afterClaimHoldings,
+    updatedTargetOutcome: afterClaimTargetOutcome,
+    exit: afterClaimExits,
+    updatedGuaranteeOutcome: afterClaimGuarantee,
+  };
+}
+
+function claimOneGuaranteeForOneAsset(
+  guarantee: Allocation,
+  targetAllocations: Allocation[],
+  holdingsForGuarantee: BigNumber,
+  exitRequest: number[]
+): {
+  afterClaimAllocations: Allocation[];
+  afterClaimGuaranteeAmount: BigNumber;
+  newExits: Allocation[];
+} {
```

Comment:
```suggestion
} {
  if (guarantee.allocationType !== AllocationType.guarantee) {
    throw Error("Expected allocation type guarantee");
  }
```

Reply:
I think I'll create a follow up PR for this.